### PR TITLE
HR-1181 Morph Documentation and Testing Template

### PR DESCRIPTION
**Pull request by TheGrandPackard**

### CODE REVIEW

File: routes/routes.go [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+// * Exchange
+// * Gift Return (for store credit)
+// * Gift Exchange
+func previewReturn(w http.ResponseWriter, r *http.Request) {
+	postReturn(w, r, false)
+}
+
+// The start return route is a Happy Returns side process where the return data is stored locally, but the return is not submitted to the retailer’s API for completion. This route may make calls to the retailer's API to start a return to generate information to use later in the return flow, depending on the capabilities of the retailer's API. The shape of the request body and response body for preview is consistent with the Preview and Return routes.
+//
+// The start return route is optional to implement, and is generally implememented because of a process requirement for how returns are processed on the retailer's end so that data is available when finishing the return.
```
Comment (tim-dev) [inline routes/routes.go]: This route is not optional for hosted retailers.  The "we don't actually do anything" version of this just returns the preview response.

File: routes/routes.go [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+// * Refund (to original payment method)
+// * Store Credit
+// * Exchange
+// * Gift Return (for store credit)
+// * Gift Exchange
+func previewReturn(w http.ResponseWriter, r *http.Request) {
+	postReturn(w, r, false)
+}
+
+// The start return route is a Happy Returns side process where the return data is stored locally, but the return is not submitted to the retailer’s API for completion. This route may make calls to the retailer's API to start a return to generate information to use later in the return flow, depending on the capabilities of the retailer's API. The shape of the request body and response body for preview is consistent with the Preview and Return routes.
```
Comment (tim-dev) [inline routes/routes.go]: The first line says it's a "Happy Returns side process" but then the other lines talk about how the retailer might need to do stuff.  I'd change it to say something like "The start return route is called when a customer starts a return online, and can be used by the retailer to create an RMA or other similar logic, if necessary"

File: docs/apib.tmpl [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+This documentation serves as an implementation guide for new morphs being written to implement morphs for new retailer platforms and custom retailer systems. There is a contract that exists between the front-end services and the morphs that must be upheld in order for the return flow to work as expected, which is detailed below. 
+
+The manner in which the morphs integrate with the retailer platform or custom system will vary depending on which APIs and web interfaces are available, and are not in the scope of this documentation. Also, the routes described below do not describe all of the necessary routes that need to be implemented by the retailer to support the Happy Returns return flow. More information about what routes are required to be implemented by retailers to support the Happy Returns return flow can be found [here](https://www.example.com).
+
+## Return Types
+
+Return types are fixed string values for each of the supported types and will be used in communication between the front-end services and the morphs. The three return types are:
+* Refund "refund"
+* Store Credit "store-credit"
+* Exchange "exhange"
```
[the remark this answers is not in the document]
Comment (TheGrandPackard) [reply docs/apib.tmpl]: Fixed

File: routes/routes.go [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+	purchaseMap   = map[string]*models.Purchase{}
+	numberPattern = regexp.MustCompile(`[^\d.-]`)
+)
+
+// The login route is used to authenticate customers as they begin a return through ORES, the Returnista app (by a representative), or at a Kiosk. The order number and zip code are used to build a customer identity to be used later on throughout the return flow. This route may or may not invoke the retailer API, depending on how customer and order information is exposed through their APIs. If the order number and zip code are not sufficient to query the orders for a customer, the handler for the login route may make calls to the retailer API in order to properly authenticate the customer and provide the necessary identity (e.g. email address) to proceed with the return flow.
+//
+// The login route is required to be implemented for the return flow to be started.
+//
+// ### Login Examples
+// * Valid gift login - 200
```
[the remark this answers is not in the document]
Comment (TheGrandPackard) [reply routes/routes.go]: Fixed

File: routes/routes.go [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+	postLogin(w, r, false)
+}
+
+// The gift login route is similar to the login route, and is used to authenticate a customer that is the recipient of a gift that is starting the return, instead of the customer that placed the original order. The order number and zip code are required for for authentication, and additional information is collected such as the email address for gift recipient (to be used later for email notifications and creating a retailer website account for the store credit to be accessed with). If the order number and zip code are not sufficient to query the orders for a customer, the handler for the login route may make calls to the retailer API in order to properly authenticate the customer and provide the necessary identity (e.g. email address) to proceed with the return flow.
+//
+// Implementing the gift login route is optional, and is only implemented for retailers that have the capability to process gift returns through their API.
+//
+// ### Gift Login Examples
+// * Valid gift login - 200
+// * Order Notnot found - 400
```
[the remark this answers is not in the document]
Comment (TheGrandPackard) [reply routes/routes.go]: Fixed

File: docs/apib.tmpl [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+
+This documentation serves as an implementation guide for new morphs being written to implement morphs for new retailer platforms and custom retailer systems. There is a contract that exists between the front-end services and the morphs that must be upheld in order for the return flow to work as expected, which is detailed below. 
+
+The manner in which the morphs integrate with the retailer platform or custom system will vary depending on which APIs and web interfaces are available, and are not in the scope of this documentation. Also, the routes described below do not describe all of the necessary routes that need to be implemented by the retailer to support the Happy Returns return flow. More information about what routes are required to be implemented by retailers to support the Happy Returns return flow can be found [here](https://www.example.com).
```
[the remark this answers is not in the document]
Comment (TheGrandPackard) [reply docs/apib.tmpl]: I was planning to refer to an additional document, but it doesn't exist yet. I changed the verbage to see the retailer's documentation instead.

File: docs/apib.tmpl [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+
+## Return Types
+
+Return types are fixed string values for each of the supported types and will be used in communication between the front-end services and the morphs. The three return types are:
```
[the remark this answers is not in the document]
Comment (TheGrandPackard) [reply docs/apib.tmpl]: Frontman/Operator are expecting fixed string values back for those types though, right? Looking at Jaanuu, there is a translation map of these strings from what I have in the documentation and what Jaanuu requires.  https://github.com/happyreturns/morph-jaanuu/blob/master/jaanuu.go#L27

File: routes/routes.go [hunk 8]
```diff
@@ -0,0 +1,390 @@
...
+		info.Email = "dev@happyreturns.com"
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	if err := json.NewEncoder(w).Encode(info); err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+	}
+}
+
+// The get purchases route is called after the customer logs in, and the customer identity is provided in order to query a list of orders from the retailer's API. After getting the list of orders, the returnability of each purchase is determined in the handler and returned to the front-end.
```
[the remark this answers is not in the document]
Comment (TheGrandPackard) [reply routes/routes.go]: Thanks for the clarification - I am finally getting a grasp of frontman vs. operator authentication and what it is useful for. I reworded this to be more accurate.

File: routes/routes.go [hunk 9]
```diff
@@ -0,0 +1,390 @@
...
+//
+// The purchases route is required to be implemented to populate the purchases for a customer after authenticating.
```
[the remark this answers is not in the document]
Comment (TheGrandPackard) [reply routes/routes.go]: Updated for accuracy.

File: routes/routes.go [hunk 10]
```diff
@@ -0,0 +1,390 @@
...
+		purchase.When = time.Now().Add(time.Second * 60 * 24 * 7 * -1).Format(time.RFC3339)
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	if err := json.NewEncoder(w).Encode(response); err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+	}
+}
+
+// The return options route returns all of the return options for a given purchase. The return options response will include the option to refund the purchase (to the original payment option), receive store credit, or exchange the purchase for another similar item of the same cost (so that there is credit card or store credit processing necessary). If the item is exchangeable, exchange options will be returned to be selected from, as well as a suggested exchange option depending on return reason, if possible (for one click exchanges). If the retailer API does not provide sufficient information along with the data returned in the purchases route to determine exchange options, that information would be queried from the retailer API at this step to build out the exhange options, attributes, and suggestions (if applicable).
```
[the remark this answers is not in the document]
Comment (TheGrandPackard) [reply routes/routes.go]: Changed.

File: routes/routes.go (hunk 1 above)
Comment (TheGrandPackard) [reply routes/routes.go]: That makes way more sense. I updated the comments to relay that information.

Comment (TheGrandPackard) [reply routes/routes.go]: Taking the inverse of your statement, is this route optional for retailers thathave return bar returns and/or mail drop-off as options, but not hosted returns?

File: routes/routes.go (hunk 2 above)
Comment (TheGrandPackard) [reply routes/routes.go]: Understood and changed.

File: routes/routes.go [hunk 11]
```diff
@@ -0,0 +1,390 @@
...
+//
+// ### Login Examples
+// * Valid gift login - 200
+// * Order not found - 400
+// * Order number and zipcode don't match - 401
+func login(w http.ResponseWriter, r *http.Request) {
+	postLogin(w, r, false)
+}
+
+// The gift login route is similar to the login route, and is used to authenticate a customer that is the recipient of a gift that is starting the return, instead of the customer that placed the original order. The order number and zip code are required for for authentication, and additional information is collected such as the email address for gift recipient (to be used later for email notifications and creating a retailer website account for the store credit to be accessed with). If the order number and zip code are not sufficient to query the orders for a customer, the handler for the login route may make calls to the retailer API in order to properly authenticate the customer and provide the necessary identity (e.g. email address) to proceed with the return flow.
```
[the remark this answers is not in the document]
Comment (TheGrandPackard) [reply routes/routes.go]: That makes sense. I updated the wording.

File: routes/routes.go [hunk 12]
```diff
@@ -0,0 +1,390 @@
...
+
+var (
+	purchases     []*models.Purchase
+	purchaseMap   = map[string]*models.Purchase{}
+	numberPattern = regexp.MustCompile(`[^\d.-]`)
+)
+
+// The login route is used to authenticate customers as they begin a return through ORES, the Returnista app (by a representative), or at a Kiosk. The order number and zip code are used to build a customer identity to be used later on throughout the return flow. This route may or may not invoke the retailer API, depending on how customer and order information is exposed through their APIs. If the order number and zip code are not sufficient to query the orders for a customer, the handler for the login route may make calls to the retailer API in order to properly authenticate the customer and provide the necessary identity (e.g. email address) to proceed with the return flow.
+//
+// The login route is required to be implemented for the return flow to be started.
```
[the remark this answers is not in the document]
Comment (TheGrandPackard) [reply routes/routes.go]: Changed.

File: routes/routes.go [hunk 13]
```diff
@@ -0,0 +1,390 @@
...
+	InvalidZipcode     = "99999"
+)
+
+var (
+	purchases     []*models.Purchase
+	purchaseMap   = map[string]*models.Purchase{}
+	numberPattern = regexp.MustCompile(`[^\d.-]`)
+)
+
+// The login route is used to authenticate customers as they begin a return through ORES, the Returnista app (by a representative), or at a Kiosk. The order number and zip code are used to build a customer identity to be used later on throughout the return flow. This route may or may not invoke the retailer API, depending on how customer and order information is exposed through their APIs. If the order number and zip code are not sufficient to query the orders for a customer, the handler for the login route may make calls to the retailer API in order to properly authenticate the customer and provide the necessary identity (e.g. email address) to proceed with the return flow.
```
[the remark this answers is not in the document]
Comment (TheGrandPackard) [reply routes/routes.go]: Understood. This has been really insightful to properly understand the architecture and how morphs are leveraged in the return flow.

File: routes/routes.go (hunk 1 above)
Comment (TheGrandPackard) [reply routes/routes.go]: I rewrote that whole comment section based off our discussion that clarified my understanding of how this process works.

File: docs/apib.tmpl (hunk 7 above)
[the remark this answers is not in the document]
Comment (TheGrandPackard) [reply docs/apib.tmpl]: Updated the wording to be more clear about how return type identifiers are fixed, but labels are arbitrary.

File: routes/routes.go [hunk 14]
```diff
@@ -0,0 +1,390 @@
...
+//
+// ### Login Examples
+// * Valid login - 200
+// * Order not found - 400
+// * Order number and zipcode don't match - 401
+func login(w http.ResponseWriter, r *http.Request) {
+	postLogin(w, r, false)
+}
+
+// The gift login route is similar to the login route, and is used to authenticate a customer that is the recipient of a gift that is starting the return, instead of the customer that placed the original order. The order number and zip code are required for for authentication, and the gift recipient's email address is collected (to be used later for email notifications and creating a retailer website account for the store credit to be accessed with). The email address of the customer that placed the order should not be returned as the identity, and the identity should remain the order number so that the gift recipient will only see the items in the order they received the gift from.
```
Comment (tim-dev) [inline routes/routes.go]: One slight addendum: in addition to the email being provided, if they use hosted we also collect the shipping address so they can do exchanges.

File: routes/routes.go [hunk 15]
```diff
@@ -0,0 +1,390 @@
...
+		info.Email = "dev@happyreturns.com"
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	if err := json.NewEncoder(w).Encode(info); err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+	}
+}
+
+// The get purchases route is called to query a list of orders relevant to the customer. The returnability of each purchase is provided with the purchases so the customer knows what can be
```
Comment (tim-dev) [inline routes/routes.go]: what can be...?

File: routes/routes.go [hunk 16]
```diff
@@ -0,0 +1,390 @@
...
+// * Refund (to original payment method)
+// * Store Credit
+// * Exchange
+// * Gift Return (for store credit)
+// * Gift Exchange
+func previewReturn(w http.ResponseWriter, r *http.Request) {
+	postReturn(w, r, false)
+}
+
+// The start return route is called when a return is started using the hosted returns (ORES) return offering, and is provided to allow retailers a way to do any custom actions after the customer has indicated that they are going to return an item at a return bar or by mail drop-off, but the item has not been received by the hub. The retailer may wish to start an RMA or perform some other action at this point in the return flow.
```
Comment (tim-dev) [inline routes/routes.go]: The first sentence is pretty long and kinda confusing.  I'd say:
 "The start return route is called when a return is started using the hosted returns (ORES) return offering.  This is provided to allow retailers a chance to do anything they'd need to do when a return is started but not received or refunded (examples: create an RMA, reserve inventory for an exchange)."

File: routes/routes.go [hunk 17]
```diff
@@ -0,0 +1,390 @@
...
+//
+// This route is only required to be implemented if the retailer has hosted returns (ORES). This route is not called for in-store returns.
+func startReturn(w http.ResponseWriter, r *http.Request) {
+	postReturn(w, r, false)
+}
+
+// The return route processes the return for completion to the retailer’s API. For exchanges, this will also process the new order for the exchanged item if the retailer API supports exchanges.
+//
+// Gift returns and gift exchanges are handled by this same route in an identical fashion, except that the return is flagged as a gift return in the customer identity that is passed through with the request.
```
Comment (tim-dev) [inline routes/routes.go]: Probably worth noting that for gift returns, the CustomerIdentity email should be used for store credit, and the address for exchanges.

File: routes/routes.go (hunk 14 above)
Comment (TheGrandPackard) [reply routes/routes.go]: Updated

File: routes/routes.go (hunk 15 above)
Comment (TheGrandPackard) [reply routes/routes.go]: Finished that thought

File: routes/routes.go (hunk 16 above)
Comment (TheGrandPackard) [reply routes/routes.go]: Copied as-is

File: routes/routes.go (hunk 17 above)
Comment (TheGrandPackard) [reply routes/routes.go]: Added

File: routes/routes.go [hunk 18]
```diff
@@ -82,7 +82,7 @@ func postLogin(w http.ResponseWriter, r *http.Request, isGift bool) {
 	}
 }
 
-// The get purchases route is called to query a list of orders relevant to the customer. The returnability of each purchase is provided with the purchases so the customer knows what can be
+// The get purchases route is called to query a list of orders relevant to the customer. The returnability of each purchase is provided with the purchases so the customer knows what can be returned, or why the purchase can no longer be returned (out of return window, already returned, final sale, etc.). The purchases route will return all purchases for a customer if possible, and only the relevant purchase for gift retutrns.
```
Comment (tim-dev) [inline routes/routes.go]: "retutrns"

Comment (TheGrandPackard) [reply routes/routes.go]: Fixed

**Pull request merged**